diff --git a/build_tools/menuselect-deps.in b/build_tools/menuselect-deps.in
index 7502553..e8405b1 100644
--- a/build_tools/menuselect-deps.in
+++ b/build_tools/menuselect-deps.in
@@ -4,6 +4,7 @@ COROSYNC=@PBX_COROSYNC@
 CRYPTO=@PBX_CRYPTO@
 BFD=@PBX_BFD@
 BISON=@PBX_BISON@
+BSON=@PBX_BSON@
 CODEC2=@PBX_CODEC2@
 CURL=@PBX_CURL@
 DAHDI=@PBX_DAHDI@
@@ -35,6 +36,7 @@ XMLSTARLET=@PBX_XMLSTARLET@
 BASH=@PBX_BASH@
 LUA=@PBX_LUA@
 MISDN=@PBX_MISDN@
+MONGOC=@PBX_MONGOC@
 MYSQLCLIENT=@PBX_MYSQLCLIENT@
 NBS=@PBX_NBS@
 NETSNMP=@PBX_NETSNMP@
diff --git a/configure.ac b/configure.ac
index 583a576..fb37bf9 100644
--- a/configure.ac
+++ b/configure.ac
@@ -575,6 +575,8 @@ AST_EXT_LIB_SETUP([TINFO], [Term Info], [tinfo])
 AST_EXT_LIB_SETUP([TONEZONE], [tonezone], [tonezone])
 AST_EXT_LIB_SETUP([UNBOUND], [unbound], [unbound])
 AST_EXT_LIB_SETUP([UNIXODBC], [unixODBC], [unixodbc])
+AST_EXT_LIB_SETUP([BSON], [bson-1.0], [bson-1.0])
+AST_EXT_LIB_SETUP([MONGOC], [mongoc-1.0], [mongoc-1.0])
 AST_EXT_LIB_SETUP([VORBIS], [Vorbis], [vorbis])
 AST_EXT_LIB_SETUP([VPB], [Voicetronix API], [vpb])
 AST_EXT_LIB_SETUP([X11], [X11], [x11])
@@ -2450,6 +2452,10 @@ if test "${PBX_SQLITE3}" != 1; then
 	exit 1
 fi
 
+# for ast_mongo
+AST_EXT_LIB_CHECK([BSON], [bson-1.0], [bson_init], [libbson-1.0/bson.h], [], [-I/usr/local/include/libbson-1.0])
+AST_EXT_LIB_CHECK([MONGOC], [mongoc-1.0], [mongoc_init], [libmongoc-1.0/mongoc.h], [], [-I/usr/local/include/libbson-1.0 -I/usr/local/include/libmongoc-1.0])
+
 # Find crypt support
 # * -lcrypt on *NIX
 # * in libsystem on OS X
diff --git a/makeopts.in b/makeopts.in
index 8168ee9..e60206a 100644
--- a/makeopts.in
+++ b/makeopts.in
@@ -335,6 +335,12 @@ UNBOUND_LIB=@UNBOUND_LIB@
 UNIXODBC_INCLUDE=@UNIXODBC_INCLUDE@
 UNIXODBC_LIB=@UNIXODBC_LIB@
 
+BSON_INCLUDE=@BSON_INCLUDE@
+BSON_LIB=@BSON_LIB@
+
+MONGOC_INCLUDE=@MONGOC_INCLUDE@
+MONGOC_LIB=@MONGOC_LIB@
+
 UUID_INCLUDE=@UUID_INCLUDE@
 UUID_LIB=@UUID_LIB@
 
